cli/listener/src/main.rs

```diff
@@ -165,46 +170,55 @@ pub fn get_interval_price_for_period(
             price_lamports_denominator: row.get(6)?,
         })
     };
+    // This is the constructor for mainnet.
+    let epoch_schedule = EpochSchedule::without_warmup();
 
     let (first, last) = {
-        // Get first logged minimal logged data based on timestamp that is greater than `from_time`.
-        // TODO(516): Do not limit the query below, but select the first data point
-        // that is offset by 200 data points from the selected epoch
-        let stmt_first = &mut tx.prepare(
-            "WITH prices_epoch AS (
-                SELECT *
-                FROM exchange_rate
-                WHERE epoch = (SELECT MIN(epoch) from exchange_rate where pool = :pool AND timestamp > :t)
-              )
-              SELECT
-                *
-              FROM
-                prices_epoch
-              ORDER BY
-                timestamp ASC
-            ",
+        // Get minimum epoch in which timestamp is greater than `from_time`.
+        let mut first_epoch_stmt = tx.prepare(
+            "SELECT MIN(epoch) from exchange_rate where pool = :pool and timestamp > :t",
+        )?;
+        let epoch = match first_epoch_stmt
+            .query_row([pool.clone(), from_time.to_rfc3339()], |row| {
+                row.get::<usize, u64>(0)
+            })
+            .ok()
+        {
+            Some(epoch) => epoch,
+            None => return Ok(None),
+        };
+        let minimum_slot = epoch_schedule.get_first_slot_in_epoch(epoch) + QUERY_SLOT_OFFSET;
+
+        // Get the first row from `epoch` in which the slot is greater than `minimum_slot`.
+        let mut exchange_rate_stmt = tx.prepare(
+            "SELECT * from exchange_rate WHERE epoch = :epoch AND slot >= :slot_min LIMIT 1",
```

Comment:
This query is missing a filter on `AND pool = :pool`. For now we only have a single pool, but if we ever add a second one, this might return a row for a different pool than the one we tried to query.